pynocular/backends/sql.py

```diff
@@ -0,0 +1,192 @@
+"""Contains the SQLDatabaseModelBackend class"""
+
+import logging
+from typing import Any, Dict, List, Optional
+
+from databases import Database
+from databases.core import Transaction
+from sqlalchemy import and_
+from sqlalchemy.dialects.postgresql import insert
+from sqlalchemy.sql.elements import BinaryExpression, UnaryExpression
+
+from pynocular.backends.base import DatabaseModelBackend, DatabaseModelConfig
+from pynocular.exceptions import InvalidFieldValue, InvalidTextRepresentation
+
+logger = logging.getLogger("pynocular")
+
+
+class SQLDatabaseModelBackend(DatabaseModelBackend):
+    """SQL database model backend
+
+    This backend works with SQL dialects supported by https://www.encode.io/databases/. except sqlite*
+
+    * sqlalchemy does not support the `RETURNING` clause. See https://github.com/sqlalchemy/sqlalchemy/issues/6195
+    """
+
+    def __init__(self, db: Database):
+        """Initialize a SQLDatabaseModelBackend
+
+        Args:
+            db: Database object that has already established a connection
+
+        """
+        self.db = db
+
+    def transaction(self) -> Transaction:
+        """Create a new transaction
+
+        Returns:
+            new transaction to be used as a context manager
+
+        """
+        return self.db.transaction()
+
+    async def select(
+        self,
+        config: DatabaseModelConfig,
+        where_expressions: Optional[List[BinaryExpression]] = None,
+        order_by: Optional[List[UnaryExpression]] = None,
+        limit: Optional[int] = None,
+    ) -> List[Dict[str, Any]]:
+        """Select a group of records
+
+        Args:
+            config: DatabaseModelConfig instance that contains references to a table and
+                columns that can be used to build queries suited to the backend.
+            where_expressions: A list of BinaryExpressions for the table that will be
+                `and`ed together for the where clause of the backend query
+            order_by: A list of criteria for the order_by clause
+            limit: The number of records to return
+
+        Returns:
+            list of records
+
+        Raises:
+            InvalidFieldValue: The class is missing a database table
+
+        """
+        async with self.transaction():
```

Comment:
are we putting every query in a transaction now?

Reply:
I just saw that and removed it from the select statement. All methods that do mutation are wrapped in a transaction. I thought we could keep it simple and not do the conditional thing, but if there is a reason to add it lmk